### PR TITLE
ci: fix clang-tidy

### LINKED ISSUE #9512: clang-tidy is broken
There are currently 2 issues with clang-tidy:
1. When clang-tidy has an error, the CI job still reports back "pass"
2. clang-tidy *always* has errors on every PR such as the following:
`error: invalid argument '-std=c++14' not allowed with 'C' [clang-diagnostic-error]`
`error: cannot use 'throw' with exceptions disabled [clang-diagnostic-error]`

I'm still troubleshooting issue 1. [This PR](https://github.com/envoyproxy/envoy/pull/9140) added a fix (`-o pipefail`) to unmask an error in `git clone`, but for some reason errors in clang-tidy itself are still being swallowed.

For issue 2 I'm suspecting that it's caused by [this change](https://github.com/envoyproxy/envoy/pull/8628/files#diff-bd94107c801983d39bc0bd34030f0b12L62) in `gen_compilation_database.py` which removes the hard-coded `clang++` compiler invocation in favor of whatever compiler the Bazel magic returns. From observing my `compile_commands.json`, it appears to be the value of `$CC`, not `$CXX` (i.e. I get `gcc` instead of `g++`) which explains why clang-tidy doesn't like exceptions or `-std=c++14`. Interestingly, `-x c++` is correctly set, although I believe to correctly compile C++ with a C compiler you also need to link the stdlib, such as `-lc++`.

cc @lizan

### PR DESCRIPTION
**Pull request by lizan**
Description:
Fixes #9512.

Risk Level:
Testing:
Docs Changes:
Release Notes:

Signed-off-by: Lizan Zhou <lizan@tetrate.io>

### CODE REVIEW

File: test/integration/fake_upstream.cc [hunk 1]
```diff
@@ -83,10 +83,10 @@ void FakeStream::encode100ContinueHeaders(const Http::ResponseHeaderMap& headers
 void FakeStream::encodeHeaders(const Http::HeaderMap& headers, bool end_stream) {
   std::shared_ptr<Http::ResponseHeaderMap> headers_copy(
       Http::createHeaderMap<Http::ResponseHeaderMapImpl>(headers));
-  if (add_served_by_header_) {
+  if (add_served_by_header_)
```
Comment (htuch) [inline test/integration/fake_upstream.cc]: What is this style change? I thought we had mandatory braces for single line `if`. I think Heartbleed or similar was caused by not doing this.

File: .clang-tidy [hunk 2]
```diff
@@ -1,19 +1,4 @@
-Checks:          'abseil-*,
-                  bugprone-*,
-                  clang-analyzer-*,
```
Comment (htuch) [inline .clang-tidy]: Are we regressing anything by eliminating these?

File: test/integration/fake_upstream.cc (hunk 1 above)
Comment (lizan) [reply test/integration/fake_upstream.cc]: No, I'm trying to prove clang-tidy will fail on this, and revert later.

File: .clang-tidy (hunk 2 above)
Comment (lizan) [reply .clang-tidy]: Those are checked but not enforced before but I don't think anyone is looking at it. So just check what we are enforcing. Then all errors are written to the fix yaml file so we can check in bash easily.

**Comment by lizan** [comment]
Need https://github.com/envoyproxy/envoy-build-tools/pull/39 and a image update for export-fixes

File: ci/run_clang_tidy.sh [hunk 3]
```diff
@@ -60,17 +61,27 @@ if [[ "${RUN_FULL_CLANG_TIDY}" == 1 ]]; then
...
       -clang-tidy-binary=${CLANG_TIDY} \
-      -p 1
+      -export-fixes=${FIX_YAML} \
+      -j ${NUM_CPUS:-0} -p 1 -quiet
+fi
+
+if [[ -s "${FIX_YAML}" ]]; then
+  echo "clang-tidy check failed, potentially fixed by clang-apply-replacements:"
+  cat ${FIX_YAML}
+  exit 1
```
Comment (derekargueta) [inline ci/run_clang_tidy.sh]: nice workaround to the non-informative exit code

**Comment by derekargueta** [comment]
If you need 2 more test cases, these should be updated to `nullptr` which is covered by `modernize-use-nullptr`
https://github.com/envoyproxy/envoy/blob/master/source/common/http/utility.cc#L555
https://github.com/envoyproxy/envoy/blob/master/source/common/api/win32/os_sys_calls_impl.cc#L58

File: .clang-tidy (hunk 2 above)
Comment (mattklein123) [reply .clang-tidy]: Once we get this fixed is it possible to do a complete master run and fix all the regressions? I'm sure there are a lot. Thanks for doing this. Very excited to get this fixed!

**Comment by marcomagdy** [comment]
FYI - I don't mind doing the busy work of fixing the things flagged by clang-tidy once we have this correctly reporting the problems. I'll merge Lizan's branch locally and have a separate PR fixing the issues that we can merge, and that would make this PR green to merge.

File: ci/run_clang_tidy.sh [hunk 4]
```diff
@@ -60,17 +61,27 @@ if [[ "${RUN_FULL_CLANG_TIDY}" == 1 ]]; then
   "${LLVM_PREFIX}/share/clang/run-clang-tidy.py" \
```
Comment (marcomagdy) [inline ci/run_clang_tidy.sh]: python3 here as well?

**Comment by lizan** [comment]
I'm waiting for image update #10506 to make this CI green

**Comment by marcomagdy** [comment]
> I'm waiting for image update #10506 to make this CI green

oh yeah - it only runs on the diff.

**Comment by lizan** [comment]
@htuch good to go, verified clang-tidy CI detects failure: https://dev.azure.com/cncf/envoy/_build/results?buildId=35761&view=logs&jobId=3bf5ec37-bb6f-5fb6-2cf1-5134dc6106d6&j=3bf5ec37-bb6f-5fb6-2cf1-5134dc6106d6&t=41d571fe-3811-578d-07de-68468f467dae

**Pull request merged**